operator/controller/dnsnameresolver/controller.go

```diff
@@ -40,74 +37,38 @@ type Config struct {
 
 // reconciler handles the actual DNSNameResolver reconciliation logic in response to events.
 type reconciler struct {
-	dnsNameResolverCache cache.Cache
-	client               client.Client
-	resolver             *Resolver
-	startResolver        sync.Once
+	config Config
+
+	cache         cache.Cache
+	client        client.Client
+	resolver      *Resolver
+	startResolver sync.Once
 }
 
 // NewUnmanaged creates and returns a controller that watches DNSNameResolver
 // objects. The controller re-resolves the DNS names which get added to the
 // status of the DNSNameResolver objects. It also removes IP addresses of DNS
-// names, whose TTLs have expired, from the status. This is an unmanaged
-// controller, which means that the manager does not start it.
-func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, []cache.Cache, error) {
-	// Create a new cache for tracking the DNSNameResolver resources in
-	// the DNSNameResolverNamespace.
-	dnsNameResolverCache, err := cache.New(mgr.GetConfig(), cache.Options{
-		Scheme: mgr.GetScheme(),
-		DefaultNamespaces: map[string]cache.Config{
-			config.DNSNameResolverNamespace: {},
-		},
-	})
-	if err != nil {
-		return nil, nil, err
-	}
-
-	// Create a new cache to track the EndpointSlices corresponding to the
-	// CoreDNS pods.
-	corednsEndpointsSliceCache, err := cache.New(mgr.GetConfig(), cache.Options{
-		Scheme: mgr.GetScheme(),
-		DefaultNamespaces: map[string]cache.Config{
-			config.OperandNamespace: {},
-		},
-		DefaultLabelSelector: labels.SelectorFromSet(labels.Set{
-			discoveryv1.LabelServiceName: config.ServiceName,
-		}),
-	})
-	if err != nil {
-		return nil, nil, err
-	}
```

Comment:
I would suggest we still use the independent caches, rather than the `mgr.GetCache()` and `mgr.GetClient()`.

If we don't use dnsNameResolverCache and enable watching of the `config.DNSNameResolverNamespace` by default by the manager's cache, then in CDO it'll always watch for the resources in that namespace even if the DNSNameResolver feature gate is not enabled.

Regarding using the client and not the cache for the coredns endpoint slice, I am a little bit sceptical about it. The minimum TTL for DNS resolution is 5 seconds and in that scenario the client will be sending requests to the api-server every 5 seconds. On the other hand the coredns pods' ips may not change that frequently. Thus it'll be more optimal to use the cache rather than the client.

PLMKyour thoughts.

Reply:
> If we don't use dnsNameResolverCache and enable watching of the config.DNSNameResolverNamespace by default by the manager's cache, then in CDO it'll always watch for the resources in that namespace even if the DNSNameResolver feature gate is not enabled.

Why creating this controller if the feature gate is disabled?

> Regarding using the client and not the cache for the coredns endpoint slice, I am a little bit sceptical about it. The minimum TTL for DNS resolution is 5 seconds and in that scenario the client will be sending requests to the api-server every 5 seconds. On the other hand the coredns pods' ips may not change that frequently. Thus it'll be more optimal to use the cache rather than the client.

The default controller-runtime's client is [a split one and it does use the cache for the reads](https://pkg.go.dev/sigs.k8s.io/controller-runtime#hdr-Clients_and_Caches).  As a matter of fact, it would be more consistent to use it everywhere.

Reply:
> > If we don't use dnsNameResolverCache and enable watching of the config.DNSNameResolverNamespace by default by the manager's cache, then in CDO it'll always watch for the resources in that namespace even if the DNSNameResolver feature gate is not enabled.
> 
> Why creating this controller if the feature gate is disabled?

We won't create the controller when the feature gate is disabled, however as we are using the manager's cache it'll anyway get the events for the `config.DNSNameResolverNamespace` resources. If we use the specific caches then only when the controller is created at that time the cache will be created in CDO.



> > Regarding using the client and not the cache for the coredns endpoint slice, I am a little bit sceptical about it. The minimum TTL for DNS resolution is 5 seconds and in that scenario the client will be sending requests to the api-server every 5 seconds. On the other hand the coredns pods' ips may not change that frequently. Thus it'll be more optimal to use the cache rather than the client.
> 
> The default controller-runtime's client is [a split one and it does use the cache for the reads](https://pkg.go.dev/sigs.k8s.io/controller-runtime#hdr-Clients_and_Caches). As a matter of fact, it would be more consistent to use it everywhere.

This will still have the issue regarding the cache getting events for the controller when it is not created.

@alebedev87 should we just stick to removing the dnsnameresolver-crd controller in this PR? We can take up changes to the dnsnameresolver controller in a different PR. We can discuss regarding the pros and cons in that PR. I  also want to have @Miciah's opinion on this matter. This is something that was previously discussed and Miciah had shared the discussion that we had here (https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/2#discussion_r1575956173).

Reply:
> We won't create the controller when the feature gate is disabled, however as we are using the manager's cache it'll anyway get the events for the config.DNSNameResolverNamespace resources.

`config.DNSNameResolverNamespace` is a namespace you can skip adding it to [the default namespaces of the manager's cache](https://github.com/openshift/cluster-dns-operator/blob/a975508a6bf7ba15e71409118b784ec030a48efc/pkg/operator/operator.go#L92-L96) if the feature gate is disabled. Then you won't create the controller therefore no `Watch` will be used, no events either.

> should we just stick to removing the dnsnameresolver-crd controller in this PR?

Then how would we resolve [the dilemma in C-D-O](https://github.com/openshift/cluster-dns-operator/pull/394/files#r1597289267)?

> I also want to have @Miciah's opinion on this matter. This is something that was previously discussed and Miciah had shared the discussion that we had here (https://github.com/openshift/coredns-ocp-dnsnameresolver/pull/2#discussion_r1575956173).

Miciah is on PTO till the end of May, we have take a decision on our own.

Reply:
> > should we just stick to removing the dnsnameresolver-crd controller in this PR?
> 
> Then how would we resolve [the dilemma in C-D-O](https://github.com/openshift/cluster-dns-operator/pull/394/files#r1597289267)?

This is the dilemma specific for the helper controller (dnsnameresolver-crd controller). It's not related to the different caches being used in dnsnameresolver controller. We can still have multiple caches as can be seen here in cluster-ingress-operator which can be added to the manager when the controller is created: https://github.com/openshift/cluster-ingress-operator/blob/5b4f6283f48046b85f60bf9d75fa5c9222329cd1/pkg/operator/controller/route-metrics/controller.go#L43-L51

Reply:
I see your point. Yes, we can follow up on the cache logic later. Let's keep this PR related to the removal of the crd controller.